core/src/main/java/kafka/server/SharePartition.java

```diff
@@ -274,63 +272,64 @@ long endOffset() {
      * @return The next fetch offset that should be fetched from the leader.
      */
     public long nextFetchOffset() {
-        if (findNextFetchOffset.compareAndSet(true, false)) {
-            lock.writeLock().lock();
-            try {
-                // Find the next fetch offset.
-                long firstOffset = nextFetchOffset;
-                // If the re-computation is required then there occurred an overlap with the existing
-                // in-flight records, hence find the batch from which the last offsets were acquired.
-                Map.Entry<Long, InFlightBatch> floorEntry = cachedState.floorEntry(nextFetchOffset);
-                if (floorEntry != null) {
-                    firstOffset = floorEntry.getKey();
+
+        // The logic to find the next offset to start fetching from depends on findNextFetchOffset flag. If this flag
+        // is set to true, then the next fetch offset should be recomputed. If the flag is false, then the next fetch
+        // offset is Share Partition End Offset + 1. The flag is set to true in the following cases -
+        // 1. When some previously acquired records are acknowledged with type RELEASE
+        // 2. When the record lock duration expires for some acquired records using the method releaseAcquisitionLockOnTimeout
+        // 3. When some records are released on final Fetch/ Acknowledge request using the method releaseAcquiredRecords
+        // The re-computation of next fetch offset is done by iterating over the cachedState and finding the first available
+        // record. If no available record is found, then the next fetch offset is set to Share Partition End Offset + 1
+        // and findNextFetchOffset flag is set to false
+        lock.writeLock().lock();
+        try {
+            // When none of the records in the cachedState are in the AVAILABLE state, findNextFetchOffset will be false
+            if (!findNextFetchOffset.get()) {
+                if (cachedState.isEmpty()) {
+                    return endOffset;
```

Comment:
Because if the cachedState is empty, both the startOffset and endOffset are set to the last offset + 1. So in this case we won't need to add the extra 1 to the endOffset